[WIP] Push-data meter support

NoraCodes opened the pull request:
This is a work in progress pull request. It needs design work.

Currently, this PR adds a column called "identifier" to the Meters table which holds a string representation of the Obvius device identifier (AcquiSuite device, and CSV column). This is not a good design; if the columns of the AcquiSuite device change, data will go to the wrong meter. Ideally, we would add a database representation of the AcquiSuite device and record which meters correspond to which channels there.

NoraCodes committed:
Add basic scaffold for Obvius push data

NoraCodes committed:
Add opt-in port binding for PostgreSQL debugging

NoraCodes committed:
Add case-insensitivity middleware for Obvius API

NoraCodes committed:
Add support for multipart Obvius API requests

NoraCodes committed:
Implement dummy Obvius config manifest generation

NoraCodes committed:
DELETE/QUASH file output for viewing

NoraCodes committed:
Add Logfile table for tracking Obvius config files

NoraCodes committed:
Add IP field to Obvius config files

NoraCodes committed:
Add time-order listing of Obvius configs per IP

NoraCodes committed:
Categorize config files by MODBUS ID and serial #

NoraCodes committed:
Add real Obvius config manifest generation

NoraCodes committed:
Add CLI for listing and purging Obvius configs

NoraCodes committed:
Use correct name for Obvius config files

NoraCodes committed:
Fix lint errors in Obvius implementation

NoraCodes committed:
Add support for non-gzipped config files

NoraCodes committed:
Add CSV demultiplexing utility

NoraCodes committed:
Implement acceptance of Obvius data logs

This implementation is provisional and not production ready.

NoraCodes committed:
Add migration for Obvius Configfiles

And propagate errors with more info out of migrations.

NoraCodes committed:
Improve logging for insertion failures from Obvius

NoraCodes committed:
Test duplicate reading behavior for Obvius.

Obvius ingestion currently does NOT modify existing data, even when
passed new, updated data for the same meter and time. This is now
tested; if it needs to be modified, that's ok, but it should be
deliberate.

NoraCodes committed:
Rename logfileTests to configfileTests.

NoraCodes committed:
Fix tests hanging for 30s after each run.

The database connection was left open, so the process would wait for
PostgreSQL/PGPromise to time out. This commit adds a root suite
mocha.after() hook so that the database is disconnected as soon as all
the tests have run.

sandeepacharya464 commented:
@NoraCodes , Everything looks good. Great work :) Just some small things:

1) Travis build is failing. So, you might want to look at it. 
2) I also thinkwe should find a way to check if everything works as expected with Obvius api when we pass in obvius data. I know we have not been good with api test and this is something we might need to consider for the project as a whole. 
3) Also, although you write a general documentation of what a chunk of code does, I think it is complex enough to have detailed documentation.

NoraCodes committed:
Abstract and document the middlewares for Obvius

This includes adding the src/server/middleware/ folder, to which all
middlewares should eventually be moved.

NoraCodes committed:
Abstract Obvius constants into util module

This involves some internal restructuring of the util module, which does
not affect the public interface.

NoraCodes committed:
Change Obvius upload faling on duplicate readings

NoraCodes committed:
Move and document Obvius API testing fixtures

NoraCodes committed:
Update Obvius test artifact header exemptions

NoraCodes committed:
Fix lint violations for Obvius

sandeepacharya464 commented:
@NoraCodes  I really like the abstraction you have done. Thank you :) . Now the travis build is still failing. Do you know why that could be occuring?

huss commented:
This did part of the work to integrate Obvius. We still need to resolve the names (I believe). This is being worked on at this time and the PR is waiting for completion or replacement.

huss commented on src/server/test/web/obvius/README.md:
```diff
@@ -0,0 +1,14 @@
+	-F 'password=password' \
+	-F 'mode=LOGFILEUPLOAD' \
+	-F 'serialnumber=mb-001' \
+	-F 'logfile=@src/server/test/db/web/obvius/mb-001.log.gz'
```
```suggestion
	-F 'logfile=@src/server/test/web/obvius/mb-001.log.gz'
```
The directory is off for the location of this file. This is mostly so that anyone working on this is aware of how to run this.

huss commented:
This work has been merged with development (as of yesterday) in the branch obvius2020 on the main OED repo. Work on this will continue from that branch. I am leaving this open for now but a new PR will be created to accept Obvius integration.

huss commented:
With the formal PR #520 that includes this work, the PR is being closed.
